telegram/telegram.py

```diff
@@ -0,0 +1,194 @@
+                headers={'Content-Type': ['application/json']}
+                )
+            content = yield r.content()
+            res = json.loads(content)
+            if r.code == http.OK and res['ok']:
+                log.info('Webhook set up on %s' % self.inbound_url)
+            else:
+                log.info('Webhook setup failed: %s' % res['description'])
+
+        # Treat page redirects as errors, since Telegram seems to redirect us
+        # when our bot token is invalid
+        except ResponseFailed as e:
+            log.info(
+                'Webhook setup failed: Invalid token (request redirected)\n%s'
+                % e
+            )
+        # In case we get a response from Telegram that isn't JSON
+        except ValueError as e:
+            log.info(
+                'Webhook setup failed: Unexpected response (expected JSON)\n%s'
+                % e
+            )
+
+    @inlineCallbacks
+    def setup_transport(self):
+        yield super(TelegramTransport, self).setup_transport()
+        config = self.get_static_config()
+        self.outbound_url = '%s%s' % (config.outbound_url.rstrip('/'),
+                                      config.bot_token)
+        self.inbound_url = config.inbound_url
+        self.bot_username = config.bot_username
+
+        yield self.setup_webhook()
+
+    @inlineCallbacks
+    def handle_raw_inbound_message(self, message_id, request):
+        update = json.load(request.content)
+        # TODO: ensure we are not receiving duplicate updates
+
+        # Telegram updates can contain objects other than messages (ignore if
+        # that is the case)
+        if 'message' not in update:
+            log.info('Inbound update does not contain a message')
+            request.finish()
+            return
+        else:
+            message = update['message']
+
+        if 'text' in message:
+            message = self.translate_inbound_message(update['message'])
+            log.info(
+                'TelegramTransport receiving inbound message from %s to %s' % (
+                    message['from_addr'], message['to_addr']))
+
+            yield self.publish_message(
+                message_id=message_id,
+                content=message['content'],
+                to_addr=message['to_addr'],
+                from_addr=message['from_addr'],
+                transport_type=self.transport_type,
+                transport_name=self.transport_name,
+            )
+        else:
+            log.info('Message is not a text message')
+
+        request.finish()
+
+    def translate_inbound_message(self, message):
+        """
+        Translates inbound Telegram message into Vumi's preferred format
+        """
+        content = message['text']
+        to_addr = self.bot_username
+        # Sender field is empty if message is sent over a Telegram channel as
+        # opposed to being directly sent to our bot (in which case the channel
+        # id is what we want)
```

Comment:
I don't know the Telegram API so I can't tell from this comment which case below (`from` or `chat`) corresponds to the message being from a channel? I'm guessing the `chat` case (because a `chat` sounds like a `channel`) but I'm not 100% sure. Could we clarify the comment a bit? Maybe just mention the word `chat` somewhere in one of the options?